doc/flakes/design.md

```diff
@@ -180,145 +174,123 @@ Similarly, a minimal `flake.nix` for Nixpkgs:
-
-  > /path/to/my/repo?ref=develop
-
-  > /path/to/my/repo?rev=e72daba8250068216d79d2aeef40d4d95aff6666
-
-* > https://<server>/<path>.tar.xz(?hash=<sri-hash>)
-
-  > file:///<path>.tar.xz(?hash=<sri-hash>)
-
-   A flake distributed as a tarball. In pure evaluation mode, an SRI
-   hash is mandatory. It exposes a `date` attribute, being the newest
-   file inside the tarball.
-
-   Example:
-
-   > https://releases.nixos.org/nixos/unstable/nixos-19.03pre167858.f2a1a4e93be/nixexprs.tar.xz
-
-   > https://releases.nixos.org/nixos/unstable/nixos-19.03pre167858.f2a1a4e93be/nixexprs.tar.xz?hash=sha256-56bbc099995ea8581ead78f22832fee7dbcb0a0b6319293d8c2d0aef5379397c
-
-Note: currently, there can be only one flake per Git repository, and
-it must be at top-level. In the future, we may want to add a field
-(e.g. `dir=<dir>`) to specify a subdirectory inside the repository.
+  - https://example.org/my/repo.git
+  - ssh://git@github.com:NixOS/nix.git
+  - git://github.com/edolstra/dwarffs.git
+  - file:///home/my-user/some-repo/some-repo.git
+  - https://releases.nixos.org/nixos/unstable/nixos-19.03pre167858.f2a1a4e93be/nixexprs.tar.xz
+  - file:///<path>.tar.xz
+
+* Local, dirty paths
+  This `FlakeRef` is the equivalent of `file://<path>` used for dirty paths.
+
+  Example: /path/to/my/repo
+
+Notes:
+- Each FlakeRef (except for the Path option) allows for a Git revision (i.e.
+  commit hash) and/or referenceo(i.e. git branch name) to be added. For
+  tarbals, an SRI hash needs to be added.
+  Examples:
+  * `"nixpkgs/release-18.09"`
+  * `github:NixOS/nixpkgs/1e9e709953e315ab004951248b186ac8e2306451`
+  * `git://github.com/edolstra/dwarffs.git?ref=flake&rev=2efca4bc9da70fb001b26c3dc858c6397d3c4817`
+  * file:///<path>.tar.xz(?hash=<sri-hash>)
+- In full pure mode, no mutable `FlakeRef`s can be used
+  * No aliases, because they need to be looked up
+  * `github` requires a specified `rev`
+  * `ssh/https/git/file` require a specified `ref` _and_ `rev`
+  * `path` is always mutable
+- Flakes don't need to be top-level, but can also reside in a subdirectory. This is shown by adding `dir=<subdir>` to the `FlakeRef`.
+  Example: `./foo?dir=bar`
 
 
 ## Flake lock files
 
-This is a JSON file named `flake.lock` that maps flake identifiers
-used in the corresponding `flake.nix` to "immutable" flake references;
-that is, flake references that contain a revision (for Git
-repositories) or a content hash (for tarballs).
+A lockfile is a JSON file named `flake.lock` which contains a forrest of
+entries mapping `FlakeRef`s to the immutable `FlakeRef` they were resolved to.
 
 Example:
 
 ```json
 {
-  "nixpkgs": "github:NixOS/nixpkgs/41c0c1bf292ea3ac3858ff393b49ca1123dbd553",
-  "foo": "https://example.org/foo.tar.xz?hash=sha256-56bbc099995ea8581ead78f22832fee7dbcb0a0b6319293d8c2d0aef5379397c"
+  "nixpkgs": {
+    "uri": "github:NixOS/nixpkgs/41c0c1bf292ea3ac3858ff393b49ca1123dbd553",
+    "content-hash": "sha256-vy2UmXQM66aS/Kn2tCtjt9RwxfBvV+nQVb5tJQFwi8E="
```

Comment:
`content-hash` -> `narHash`. Also, there are a copy of other fields (`id`, `inputs`, `nonFlakeInputs`).